File: python/2127. Maximum Employees to Be Invited to a Meeting.py
class Solution:
    def maximumInvitations(self, favorite: List[int]) -> int:
        n = len(favorite)
        reversed_graph = [[] for _ in range(n)]
        for p in range(n):
            reversed_graph[favorite[p]].append(p)

        def _bfs(node: int, visited: set) -> int:
            q = deque([(node, 0)])
            max_distance = 0
            while q:
                current, distance = q.popleft()
                for neighbour in reversed_graph[current]:
                    if neighbour in visited: continue
                    visited.add(neighbour)
                    q.append((neighbour, distance + 1))
                    max_distance = max(max_distance, distance + 1)
            return max_distance

        longest_cycle, two_cycle = 0, 0
        visited = [False] * n
        for p in range(n):
            if visited[p]: continue
            distances = {}
            current = p
            distance = 0
            while True:
                if visited[current]: break
                visited[current] = True
                distances[current] = distance
                distance += 1
                neighbour = favorite[current]
                if neighbour in distances:
                    cycle_length = distance - distances[neighbour]
                    longest_cycle = max(longest_cycle, cycle_length)
                    if cycle_length == 2:
                        visited_nodes = {current, neighbour}
                        two_cycle += 2 + _bfs(neighbour, visited_nodes) + _bfs(current, visited_nodes)
                    break
                current = neighbour
        return max(longest_cycle, two_cycle)

File: java/2127. Maximum Employees to Be Invited to a Meeting.java
class Solution {
    public int maximumInvitations(int[] favorite) {
        int n = favorite.length;
        Map<Integer, List<Integer>> reversedGraph = new HashMap<>();
        for (int i = 0; i < n; i++) {
            reversedGraph.computeIfAbsent(favorite[i], k -> new ArrayList<>()).add(i);
        }
        int longestCycle = 0, twoCycle = 0;
        boolean[] visited = new boolean[n];
        for (int i = 0; i < n; i++) {
            if (visited[i]) continue;
            Map<Integer, Integer> distances = new HashMap<>();
            int current = i;
            int distance = 0;
            while (true) {
                if (visited[current]) break;
                visited[current] = true;
                distances.put(current, distance++);
                int neighbour = favorite[current];
                if (distances.containsKey(neighbour)) {
                    int cycleLength = distance - distances.get(neighbour);
                    longestCycle = Math.max(longestCycle, cycleLength);
                    if (cycleLength == 2) {
                        boolean[] cycleNodes = new boolean[n];
                        cycleNodes[current] = true;
                        cycleNodes[neighbour] = true;
                        twoCycle += 2 + bfs(current, cycleNodes, reversedGraph) + bfs(neighbour, cycleNodes, reversedGraph);
                    }
                    break;
                }
                current = neighbour;
            }
        }
        return Math.max(longestCycle, twoCycle);
    }
    private int bfs(int node, boolean[] visited, Map<Integer, List<Integer>> graph) {
        Deque<int[]> q = new LinkedList<>();
        q.addLast(new int[] {node, 0});
        int maxDistance = 0;
        while (!q.isEmpty()) {
            int[] current = q.pollFirst();
            if (!graph.containsKey(current[0])) continue;
            for (int next : graph.get(current[0])) {
                if (visited[next]) continue;
                visited[next] = true;
                maxDistance = Math.max(maxDistance, current[1] + 1);
                q.addLast(new int[] {next, current[1] + 1});
            }
        }
        return maxDistance;
    }
}
